[Hardknott] Refactor and cleanup images/ directory (part 2)

amstewart opened the pull request:
This patchset continues the refactoring changes started in meta-nilrt #307. The changes in this PR are more substantial and I expect will slightly change the artifacts that we build, though not in a way which breaks downstream dependers.

Expected Changes:
* make more images inherit `nilrt-core-image` and consolidate common `core-image` settings into that file.
* remove ARM and nilrt-nxg content from includes files.
* stop padding the rootfs archives with empty space (@gratian )
* make `nilrt-safemode-initramfs` include `nilrt-image-base` and consolidate common bootable image logic into that file.
* homogenize the image recipe `DESCRIPTION`s and generally update the README.
* Add a `nilrt-image-initramfs.inc` file with common initramfs logic from the safemode and runmode initramfs recipes.
* Deprecate the unused `ext2` image artifact type.
* Rename `nilrt-recovery-image` to `nilrt-recovery-media`.
* Rename `nilrt-runmode-system-image` to `nilrt-base-system-image`.

A few more changes are likely to come after this; but this is a good checkpoint.

# Testing
Testing in progress. 

@ni/rtos

amstewart committed:
nilrt-recovery-image: include from nilrt-core-image

The nilrt-recovery-image is abnormal in that it is a NILRT image, which
includes from OE's core-image class, but yet it does not use the
nilrt-core-image includes.

Include nilrt-core-image in nilrt-recovery-image, to make it more
consistent with the other "core" images.

The substantive changes to recovery image configuration as a result of
this change are that an ext2 image fs type is now created, and the
kernel bzImage is now realized (instead of being a symlink).

Signed-off-by: Alex Stewart <alex.stewart@ni.com>

amstewart committed:
nilrt-image-base: remove ARM and nilrt-nxg content

The ARM architecture and nilrt-nxg distro flavor are not valid for the
hardknott mainline. Remove them for cleanliness.

Signed-off-by: Alex Stewart <alex.stewart@ni.com>

amstewart committed:
nilrt-image-base.inc: do not pad the rootfs

500 MB of "IMAGE_ROOTFS_EXTRA_SPACE" was added to the
niconsole-image.inc when it was being used to build ext2 archives
for the LVComms product line. Contemporarily, this value was only being
inherited by the x64 runmode rootfs image, where we have elastic disk
space requirements.

Since this requirement is no longer needed, remove it from the common
includes.

Signed-off-by: Alex Stewart <alex.stewart@ni.com>

amstewart committed:
nilrt-safemode-initramfs: use nilrt-image-base

Much of the logic in nilrt-safemode-initramfs.bb is taken from
nilrt-image-base.inc. So we may as well just inherit from that common
file.

The install_additional_feeds function from nilrt-image-base is the only
bit which is inappropriate for an initramfs, so remove it from the
POSTPROCESS commands explicitly.

Signed-off-by: Alex Stewart <alex.stewart@ni.com>

amstewart committed:
images: deprecate host-toolchain-sysroot.bb

The host-toochain-sysroot image recipe has not been used in many NILRT
releases and is not supported in the hardknott mainline.

Remove it for cleanliness.

Signed-off-by: Alex Stewart <alex.stewart@ni.com>

amstewart committed:
images: homogenize image recipe DESCRIPTIONs

Reword the image recipe DESCRIPTION variables to make them homogenous
across the modern recipes.

Signed-off-by: Alex Stewart <alex.stewart@ni.com>

amstewart committed:
images: segregate deprecated rauc images/files content

The :files/nilrt-bundle-hooks.sh file is only meaningful in the context
of a RAUC image bundle.

Move it from the modern :files/ directory into the :rauc/ directory with
the rest of the RAUC content for cleanliness.

Signed-off-by: Alex Stewart <alex.stewart@ni.com>

amstewart committed:
images/includes: add nilrt-image-initramfs.inc

Some logic is common between runmode and safemode initramfs recipe
implementations.

For maintainability, move that common code into a
`nilrt-image-initramfs.inc` file and include it in the initramfs images.

Signed-off-by: Alex Stewart <alex.stewart@ni.com>

amstewart committed:
images/README: clarify includes mixins

Some RTOS devs expressed that it is unclear that the 'includes/' section
of the README declares mixin variable letters, which 'Images' recipes
*include*.

Modify the mixin variables to make it more clear that each one refers to
the includes file on its line.

Signed-off-by: Alex Stewart <alex.stewart@ni.com>

amstewart committed:
images/README: remove bullets from recipe entries

Since we removed the hierarchy structure from the images README, the
bullets (*) no longer have meaning in the README.

Remove them for clarity.

Signed-off-by: Alex Stewart <alex.stewart@ni.com>

amstewart committed:
nilrt-core-image.inc: remove ext2 FSTYPE

The ext2 IMAGE_FSTYPE was only used by the nilrt-nxg
minimal-nilrt-image.bb recipe, which has been long-deprecated.

Remove it to reduce the number of build steps and artifacts.

Signed-off-by: Alex Stewart <alex.stewart@ni.com>

amstewart committed:
images/README: add comments for some recipes

In review, RTOS devs asked that some of the top-level recipes receive
more descriptive comments, based on their current recipe DESCRIPTIONs.

Add them to the README.

Signed-off-by: Alex Stewart <alex.stewart@ni.com>

amstewart committed:
images: rename nilrt-recovery-image to nilrt-recovery-media

We have generally removed the `-image` suffix from image recipes because
it is not descriptive about the form of the image.

Rename 'nilrt-recovery-image' to 'nilrt-recovery-media' to describe its
form more precisely.

Signed-off-by: Alex Stewart <alex.stewart@ni.com>

amstewart committed:
images: rename nilrt-runmode-system-image

The MAX and NIPM installer stack prefers to call the NILRT runmode
rootfs, postinst, and CDF bundle: the "base system image".

Rename 'nilrt-runmode-system-image' to 'nilrt-base-system-image' to make
it clear that the recipe only exists to provide the BSI installer
metadata and supporting files.

Signed-off-by: Alex Stewart <alex.stewart@ni.com>

shruthi-ravi commented on recipes-core/images/nilrt-recovery-image.bb:
```diff
@@ -1,6 +1,6 @@
 DESCRIPTION = "Tiny initramfs image intended to run recovery and install operations for NILinux RT"
 
-IMAGE_FSTYPES = "${INITRAMFS_FSTYPES} tar.bz2 wic"
+IMAGE_FSTYPES = "wic"
```
Do you need to append here instead of assigning a value?

amstewart replied:
I do. The test build I was running just failed because of this bug. :)

shruthi-ravi commented on recipes-core/images/nilrt-recovery-image.bb:
```diff
@@ -48,4 +48,4 @@ ROOTFS_POSTPROCESS_COMMAND += "symlink_iso;"
 IMAGE_PREPROCESS_COMMAND += " bootimg_fixup; "
 
-inherit core-image
+require includes/nilrt-core-image.inc
```
I'd replace `inherit nilrt-core-image` in the commit message to `include nilrt-core-image`.

amstewart replied:
Will fix.

shruthi-ravi commented on recipes-core/images/nilrt-safemode-initramfs.bb:
```diff
@@ -23,8 +24,6 @@ IMAGE_INSTALL_NODEPS += "\
 	${NI_PROPRIETARY_SAFEMODE_PACKAGES} \
 "
 
-BAD_RECOMMENDATIONS_append_pn-${PN} += "shared-mime-info"
```
This initramfs does not exclude `ca-certificates`, but it looks like with your change, it will. Is that intended? Not sure why we exclude the certificates from the runmode initramfs.

amstewart replied:
[Ionel thought they were inappropriate]([https://github.com/ni/meta-nilrt/commit/243fdb0528ae3559e39dfac27cd29ad0f2291a7d#diff-99cc1ba30ac7f8cf446cf4217a17416d1d3a6e3c24d2911a7f424a42bf2017f0). :shrug: It's a `BAD_RECOMMENDS`, so it won't block opkg from installing `ca-certificates` if a requested package actually depends on it.

```
(safemode) admin@NI-cRIO-903x-VM-a18e5030:~# opkg whatdepends -A ca-certificates
Root set:
  ca-certificates
What depends on root set
        ca-certificates-dev 20210119-r0.75      depends on ca-certificates (= 20210119-r0.75)
        docker-ce v19.03.15-ce+git99e3ed89195c4e551e87aad1e7453b65456b03ad-r0.0 depends on ca-certificates
        docker-ce-dev v19.03.15-ce+git99e3ed89195c4e551e87aad1e7453b65456b03ad-r0.0     depends on docker-ce (= v19.03.15-ce+git99e3ed89195c4e551e87aad1e7453b65456b03ad-r0.0)
        ni-ca-certs 21.3.0.49152-0+f0   depends on ca-certificates
(safemode) admin@NI-cRIO-903x-VM-a18e5030:~# opkg whatrecommends -A ca-certificates
Root set:
  ca-certificates
What recommends root set
        wget 1.21.1-r0.0        recommends ca-certificates
        python3-crypt 3.9.5-r0.9        recommends ca-certificates
        libcurl4 7.75.0-r0.20   recommends ca-certificates
```

It seems like it gets installed to safemode anyway via the `ni-ca-certs` dependency.

gratian replied:
I'm guessing Ionel excluded them because it is non-trivial to update/revoke a certificate when it's built-in the initramfs i.e. you can't do it by just updating a package via opkg.

shruthi-ravi commented on recipes-core/images/nilrt-safemode-initramfs.bb:
```diff
@@ -23,8 +24,6 @@ IMAGE_INSTALL_NODEPS += "\
-
 # Do not allow python to be installed into safemode ramdisk due to size
 PACKAGE_EXCLUDE += "python-core python3-core"
 
```
There's an IMAGE_FSTYPES assignment down below that you may need to remove.

amstewart replied:
No; the `IMAGE_FSTYPE = "cpio.xz"` assignment is correct for the safemode initramfs. It gets installed to the `nilrt-safemode-rootfs` as `runmode.xz`.

shruthi-ravi replied:
Then, there is no need for the IMAGE_FSTYPES assignment from the `nilrt-image-initramfs.inc` includes file, and even less need for the includes file itself.

shruthi-ravi commented on recipes-core/images/includes/nilrt-image-initramfs.inc:
```diff
@@ -0,0 +1,10 @@
+# packed as a CPIO archive; and expected to provide an initial ramfs for
+# booting a later, complete rootfs.
+
+IMAGE_FSTYPES += " ${INITRAMFS_FSTYPES}"
```
Just a thought. To me, it seems like there isn't enough here to justify an includes file.

amstewart replied:
You might be right; especially after we remove the runmode initramfs. I'll consider removing it.

shruthi-ravi commented on scripts/lib/wic/canned-wks/nilrt-recovery-media.x64.wks:
```diff
@@ -2,7 +2,7 @@
 # long-description: Creates an EFI and legacy bootable hybrid ISO image
 # which can be used on optical media as well as USB media.
 
-part /boot --source isoimage-isohybrid --sourceparams="esp_label=NIRECOVERY,esp_extra_blocks=5000,loader=grub-efi,image_name=nilrt-recovery-image.x64" --ondisk cd --label NIRECOVERY-CD
```
Is there a client that consumes this image that needs to be notified?
In addition to that, you will also need to fix the build VM script as well as our ATS [configuration](https://codesearch.natinst.com/search/text?q=nilrt-recovery-image&fold_case=auto&regex=false&context=true).

amstewart replied:
I have an ni-central PR which is fixing up the component references. I got the installers, but thanks for pointing out the ATS references. I'll switch those as well.

shruthi-ravi commented on recipes-core/images/README:
```diff
@@ -9,10 +9,10 @@ c     = nilrt-core-image.inc       # NI-specific modifications to the core-image
 
 ## Images
-c     | nilrt-recovery-media.bb        # NILRT installation/recovery media ISO
+      | nilrt-base-system-image.bb  # NILRT base system image archive and CDF
```
I assume azdo installer changes will follow.

amstewart replied:
Correct.

amstewart committed:
nilrt-recovery-media: append ISO IMAGE_FSTYPES

Any strong assignment to the IMAGE_FSTYPES variable in
nilrt-recovery-media is overwritten at the end of the recipe, when it
includes the `nilrt-core-image.inc` file. As a result, 'wic' is not a
final FSTYPE.

Instead, _append the 'wic' string, so that it is properly added to the
IMAGE_FSTYPES list.

Signed-off-by: Alex Stewart <alex.stewart@ni.com>

amstewart commented:
# Patch V2
* `s/inherit/include` in the first commit message. @shruthi-ravi 
* `_append` the `IMAGE_FSTYPES` in `nilrt-recovery-media.bb`, so that the variable doesn't get messed up. Assignment now looks like this... @shruthi-ravi 
```
# $IMAGE_FSTYPES [5 operations]
#   set? /home/usr0/fast/nilrt-hardknott/sources/openembedded-core/meta/conf/machine/include/x86-base.inc:11
#     "wic"
#   set /home/usr0/fast/nilrt-hardknott/sources/openembedded-core/meta/conf/documentation.conf:215
#     [doc] "Formats of root filesystem images that you want to have created."
#   set? /home/usr0/fast/nilrt-hardknott/sources/openembedded-core/meta/conf/bitbake.conf:780
#     "tar.gz"
#   _append /home/usr0/fast/nilrt-hardknott/sources/meta-nilrt/recipes-core/images/nilrt-recovery-media.bb:3
#     " wic"
#   set /home/usr0/fast/nilrt-hardknott/sources/meta-nilrt/recipes-core/images/includes/nilrt-core-image.inc:4
#     "tar.bz2"
# pre-expansion value:
#   "tar.bz2 wic"
IMAGE_FSTYPES="tar.bz2 wic"
```

amstewart committed:
Revert "images/includes: add nilrt-image-initramfs.inc"

This reverts commit 3d8622002647c2d645af99f39572ae7442598a47.

There isn't enough shared code between the safemode and runmode
initramfs recipes to justify having a shared includes file between them.
Further, we are planning to deprecat the runmode initramfs entirely, so
the nilrt-image-initramfs.inc file would only be serving a single image
recipe.

Signed-off-by: Alex Stewart <alex.stewart@ni.com>

amstewart committed:
images: do not make CPIO images for all recipes

As written, all image recipes which inherit from the
nilrt-core-image.inc file will try to create artifacts of type
`${INITRAMFS_TYPES}`; usually, that means a 'cpio.gz' archive. The
safemode-initramfs recipe already overwrites this variable wholesale and
replaces it with 'cpio.xz' - which is the type expected by the safemode
boot logic. The runmode initramfs *does* use the 'cpio.gz' type in its
boot logic, but there is no need to foist that type upon all images.

Move the 'cpio.gz' type definition into the runmode-initramfs.

Signed-off-by: Alex Stewart <alex.stewart@ni.com>

amstewart commented:
# Patch V3
* Reverted the `nilrt-image-initramfs.inc` file out of this PR. There isn't enough common code between the initramfs images to justify it. @shruthi-ravi 
* Because I did the above, I then removed the `${INITRAMFS_FSTYPES}` variable from the core image `IMAGE_FSTYPES` - since it was adding the `cpio.gz` type to a bunch of images recipes, but only the runmode-initramfs actually used it. I moved the type assignment into the `nilrt-runmode-initramfs.bb` file itself.

amstewart commented:
Could you close any issues that you consider resolved @shruthi-ravi .

amstewart commented:
# Testing

Compared the hardknott pipeline artifacts created from this patchset against those of 21.8.0d135. The following are the differences...

``` diff
# nilrt-runmode-system-image-x64.tar -> nilrt-base-system-image-x64.tar
++ adds some salt components from sravichas recent salt_3000.2 recipe addition  # expected
-- some python3.9 differences (also salt-related)  # expected
-- lacks libinput  # OK; salt-related, I think
/boot/tmp/runmode/runmode_initramfs.gz:
	-- /etc/ca-certificates  # expected
	-- /etc/ssl/certs/*  # expected
	-- /usr/bin/openssl  # OK; I think this drops out due to the ca-certificates

# standard_x64_safemode.tar.gz:
/ramdisk.xz
	-- /usr/bin/update-mime-database  # expected
	-- /usr/share/mime  # expected
```

Also booted the provisioned QEMU VM and verified that it boots and opkg works.
